mlir/test/Catalyst/DetensorizeSCFTest.mlir

```diff
@@ -0,0 +1,105 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// RUN: quantum-opt --detensorize-scf --canonicalize --split-input-file %s | FileCheck %s
+
+// CHECK-LABEL: @test0
+module @test0 {
+  // CHECK-LABEL: @test_while_loop
```

Comment:
@dime10 Before I go ahead and  test across the board, is that a good MLIR test?

Reply:
I would probably say that the test is a bit too literal. These best practices are a good starting point:
https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

But generally you don't want to check the entire IR verbatim, because that makes for brittle tests. For example, if one dialect changes how things are printed, the SSA values happen to get different names, or maybe some operations are reordered, we want to avoid writing tests that are sensitive to these sort of changes.

The best practice is thus to
- focus on the thing being tested: only include instructions relevant to your pass
- omit parts of the IR if they are not relevant: e.g. type info (unless important), arguments, return values (e.g. no need to include `{{.*}} = ` for the return values if you are not checking them), etc., but this requires some judgement
- loosen the CHECK commands: from weak to strong (CHECK-DAG, CHECK, CHECK-NEXT). CHECK-NEXT is useful if you want to guarantee no other instruction is between this and the previous one, but otherwise probably overkill, CHECK allows intermediate instructions but still enforces an ordering, CHECK-DAG allows for lines to appear in a different order than written (useful for constant declarations for example (if they need to be included at all), it is unlikely that we care about their ordering and the canonicalizer may move them around anyway)

In order to tighten and make tests more effective, you can make use of the following:
- use variables to test relationships between operations: you might want to check that a value defined by a specific operation is used as the argument to another op instead of just any value. `[[VAR:.+]] =` to capture a regex pattern as a variable, `[[VAR]]` to use the captured text.
- verify that something is _not_ present: CHECK-NOT can be used to test that a string is no longer present, which should be useful in this case to verify the tensor version of the operation is gone